### ISSUE TITLE
Other model not available for User.on('attached'...

### ISSUE THREAD

**Issue by johnsoftek** [strongloop/loopback#411]
(LoopBack 2.0.1)

In my common/models/user.js file, I want to access a settings table to retrieve SMTP parameters. As per the docs:

```
module.exports = function(User) {

  User.on('attached', function() {
    var app = User.app,
      models = app.models

console.log(models)

   var Settings = models.Settings  // Settings is null
```

The Settings property is absent from app.models, as confirmed by the console log. The only models in app.models are "user" and "User".

If I  wrap the access to models.Settings in a setTimeout(fn, 5000), it works. So the app object is available when the 'attached' event is fired but not all models are available.

**Comment by ericprieto** [comment]
@johnsoftek how did you solve this?

**Comment by johnsoftek** [comment]
@ericprieto A bit hacky... I guess it is possible that both models are attached when the first model js is called, so you may have to take steps to ensure the action takes place only once.

In settings.js:
```javascript
  Settings.on('attached', function() {
    var app = Settings.app,
      User = app.models.user
    if (User) {
      setup_user_smtp_transport()
    }
  })
```

In user.js:
```javascript
  User.on('attached', function () {
    var app = User.app,
      Settings = app.models.settings
    if (Settings) {
      Settings.setup_user_smtp_transport()
    }
  }
```

**Comment by fabien** [comment]
@raymondfeng I would like to suggest a `boot` event or something similar, which is emitted once all models have loaded, right before any bootscripts are called. I did this in my own apps already, because it's easy to trip over the load-order issue when you have many models.

**Comment by ericprieto** [comment]
@raymondfeng I used `loopback.getModel('User')` instead of `app.models.User`. I think It works indepentent of load order.

**Comment by raymondfeng** [comment]
@ericprieto You can add a bootscript such as /server/boot/setup-user-model.js to further set up the model. The boot script will receive the `app` instance.

**Comment by fabien** [comment]
@raymondfeng I would still like an event for this - boot scripts decouple from the logic that usually defined in `models/*.(js|json)` files, I think.

**Comment by ebarault** [comment]
@fabien: would you mind sharing how you achieve this kind of boot event or your own?

@raymondfeng : other than that, i see there is a started event which emits when all boot scripts are executed (and hence every models are loaded) but I don't see any doc on how to use this. Would you have any examples at hand ?

**Comment by jesperbruunhansen** [comment]
This helped me fixing the problem!
```
module.exports = function(Model, options) {
  var app = require('../../server/server');

  var AnotherModel = app.loopback.getModel("AnotherModel");
  AnotherModel.on("attached", function(){
    //AnotherModel is initialized
  });

};
```

**Comment by ebarault** [comment]
@jesperbruunhansen as a best practive i usually use this to declare only once other models per model.js file:

```
var app = require('../../server.js');
var models = app.models;

var AnotherModel;

app.on('started', function () {
    AnotherModel     = models.AnotherModel;
});
```

**Comment by jesperbruunhansen** [comment]
Thanks @ebarault, I wasn't aware of the `startet` event - I will definitely look into your method!

**Comment by charlie-s** [comment]
This also works:

```
module.exports = function(Apples) {

    Apples.on('attached', function(ctx, next) {

        var Bananas = ctx.loopback.getModel('Bananas');

        Bananas.observe('after save', function(ctx, next) {

            console.log('A Banana was saved. Eat some Apples.');
            
            next();
        });

    });

};
```

**Comment by ebarault** [comment]
just a warning though @charlie-s : by doing `loopback.getModel()` you are using loopback's global registry. This might lead to unexpected results if you set up the app to use local registries.

**Comment by ewrayjohnson** [comment]
Does anyone have a 100% surefire solution for this?
`UnhandledPromiseRejectionWarning: error: relation "_aModel.aRelation_" does not exist `
I have tried all of the above solutions without success.  I have tried app events _booted_ and _started_.  I have tried all of the above solutions.  I have tried nesting multiple once attached.  I want an event that tells me when automigrate/autoupdate has completed and ALL models are attached to their datasources and ready to be used.
